src/theory/strings/rewrites-regexp-membership

```diff
@@ -25,175 +25,6 @@
-  (and (= (str.len s) (str.len t)) (not (= s t)))
-  (str.in_re s (re.++ r (str.to_re (str.++ sr1 sr2 t))))
-  false)
-
-(define-rule* str-in-re-strip-prefix-base-rev ((s String) (s1 String) (s2 String :list) (sr1 String) (sr2 String :list))
-  (str.in_re (str.++ s1 s2 s) (str.to_re (str.++ sr1 sr2 s)))
-  (str.in_re (str.++ s1 s2) (str.to_re (str.++ sr1 sr2)))
-  _)
-
-(define-cond-rule str-in-re-strip-prefix-base-neg-rev ((s String) (t String) (s1 String) (s2 String :list) (sr1 String) (sr2 String :list))
-  (and (= (str.len s) (str.len t)) (not (= s t)))
-  (str.in_re (str.++ s1 s2 s) (str.to_re (str.++ sr1 sr2 t)))
-  false)
-
-(define-rule str-in-re-strip-prefix-base-s-single-rev ((s String) (sr1 String) (sr2 String :list))
-  (str.in_re s (str.to_re (str.++ sr1 sr2 s)))
-  (str.in_re "" (str.to_re (str.++ sr1 sr2))))
-
-(define-cond-rule str-in-re-strip-prefix-base-s-single-neg-rev ((s String) (t String) (sr1 String) (sr2 String :list))
-  (and (= (str.len s) (str.len t)) (not (= s t)))
-  (str.in_re s (str.to_re (str.++ sr1 sr2 t)))
-  false)
-
-(define-cond-rule str-in-re-strip-char-rev ((s String) (s1 String) (s2 String :list) (r1 RegLan) (r2 RegLan :list))
-  (= (str.len s) 1)
-  (str.in_re (str.++ s1 s2 s) (re.++ r1 r2 re.allchar))
-  (str.in_re (str.++ s1 s2) (re.++ r1 r2)))
-
-(define-cond-rule str-in-re-strip-char-s-single-rev ((s String) (r1 RegLan) (r2 RegLan :list))
-  (= (str.len s) 1)
-  (str.in_re s (re.++ r1 r2 re.allchar))
-  (str.in_re "" (re.++ r1 r2)))
-
-(define-cond-rule str-in-re-req-unfold ((s String) (r1 RegLan) (r2 RegLan) (r3 RegLan :list))
-  (not (str.in_re s (re.++ r2 r3)))
-  (str.in_re s (re.++ (re.* r1) r2 r3))
-  (str.in_re s (re.++ r1 (re.* r1) r2 r3)))
-
-(define-cond-rule str-in-re-req-unfold-rev ((s String) (r1 RegLan) (r2 RegLan :list) (r3 RegLan))
-  (not (str.in_re s (re.++ r1 r2)))
-  (str.in_re s (re.++ r1 r2 (re.* r3)))
-  (str.in_re s (re.++ r1 r2 (re.* r3) r3)))
-
-(define-cond-rule str-in-re-skip-unfold ((s String) (r1 RegLan) (r2 RegLan) (r3 RegLan :list))
-  (not (str.in_re s (re.++ r1 (re.* r1) r2 r3)))
-  (str.in_re s (re.++ (re.* r1) r2 r3))
-  (str.in_re s (re.++ r2 r3)))
-
-(define-cond-rule str-in-re-skip-unfold-rev ((s String) (r1 RegLan) (r2 RegLan :list) (r3 RegLan))
-  (not (str.in_re s (re.++ r1 r2 (re.* r3) r3)))
-  (str.in_re s (re.++ r1 r2 (re.* r3)))
-  (str.in_re s (re.++ r1 r2)))
-
-(define-cond-rule str-in-re-test-unfold ((s String) (r1 RegLan) (r2 RegLan) (r3 RegLan :list))
-  (and (not (str.in_re s (re.++ r2 r3))) (not (str.in_re s (re.++ r1 (re.* r1) r2 r3))))
-  (str.in_re s (re.++ (re.* r1) r2 r3))
-  false)
-
-(define-cond-rule str-in-re-test-unfold-rev ((s String) (r1 RegLan) (r2 RegLan :list) (r3 RegLan))
-  (and (not (str.in_re s (re.++ r1 r2))) (not (str.in_re s (re.++ r1 r2 (re.* r3) r3))))
-  (str.in_re s (re.++ r1 r2 (re.* r3)))
-  false)
-
-(define-rule* str-in-re-concat-emp ((s String) (s1 String) (s2 String :list))
```

Comment:
I don't get how this is subsumed since the new rule does not use an and:

```
(define-rule re-in-cstring ((t String) (s String)) (str.in_re t (str.to_re s)) (= t s))
```

So for the string `(str.++ "" "")` the original rule would be match: `(str.in_re (str.++ "" "") (str.to_re ""))` and target: `(and (= "" "") (= "" ""))`?

For the new one it would be the same match but target: ` (= (str.++ "" "") "")`

Reply:
Yes, I omitted a detail here. The other rule is `MACRO_STR_EQ_LEN_UNIFY_PREFIX`, which elaborates into further RARE rules.  I'll update the description of the PR.